### PR TITLE
mgr/dashboard: Add Pool update endpoint 

### PR DESCRIPTION
**Pull request by sebastian-philipp**
This adds `PUT` requests to `/api/pool/<pool_name>` allowing to change existing pools. 

* Copied the fix for `qa/tasks/mgr/dashboard/helper.py` from #21066 
* Added minor tweak to `_validate_json()` to allow skipping `JLeaf()`
* Made use of `_validate_json()` in the Pool test.

### CODE REVIEW

File: src/pybind/mgr/dashboard/controllers/pool.py [hunk 1]
```diff
@@ -54,27 +58,51 @@ def list(self, attrs=None, stats=False):
...
+            else:
+                original_app_metadata = set()
+
+            for app in original_app_metadata - set(application_metadata):
+                set_app('disable', app)
+            for app in set(application_metadata) - original_app_metadata:
+                set_app('enable', app)
 
         for key, value in kwargs.items():
             CephService.send_command('mon', 'osd pool set', pool=pool, var=key, val=value)
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/controllers/pool.py]: In this line you should make sure that value is a string. Compression required ratio will be parsed as float but should be a string.

**Comment by Devp00l** [comment]
Hi changed a view more things, [here](https://paste.opensuse.org/6140030) you can find the diff of my current changes. The only thing I couldn't fix so far is that you can't disable all applications, one application will always be enabled at the end.

**Comment by Devp00l** [comment]
The same error is still there. [This](https://paste.opensuse.org/62308419) still fixes the issues.

**Comment by sebastian-philipp** [comment]
TODO: Integrate: https://github.com/ceph/ceph/pull/21614/commits/6599431cd09493d245bb8e5f176b97c76edff2fc

**Comment by tspmelo** [comment]
jenkins retest this please

**Comment by s0nea** [comment]
Hey @sebastian-philipp have you seen this issue in the Jenkins build?
```114/153 Test   #1: run-rbd-unit-tests.sh ...................***Failed``` But I'm not really sure if it's related to your PR.

**Comment by rjfd** [comment]
@sebastian-philipp a few months ago we discuss some implementations details about the pool management operations with respect to its asynchronous nature in https://github.com/ceph/ceph/pull/20870#issuecomment-374921515.

Most pool operations rely on the `send_command` operation, which is truly asynchronous (no thread wrapping is required), and the tasks could leverage this, instead of wrapping an already asynchronous command inside new a thread. But in this PR we are not taking advantage of the asynchronous implementation of `send_command`.

Another detail that I think could be addressed in this PR is the fact that the pool creation operation is not waiting for the PGs to became `clean` as we also discussed previously.

In https://github.com/rjfd/ceph/blob/wip-dashboard-pr-20865/src/pybind/mgr/dashboard/controllers/pool.py you can find an example of the pool creation operation that takes advantage of the asynchronous `send_command` operation, and also waits for the PGs stabilization. You can use this example to as the basis for the implementation of the remaining pool operations.

**Comment by sebastian-philipp** [comment]
TIL that here on GitHub, you just have to select some text in a post and press `r` in order to [quote](https://blog.github.com/2013-02-04-quick-quotes/) it. Interestingly though, markdown formatting is gone. 

> @sebastian-philipp a few months ago we discuss some implementations details about the pool management operations with respect to its asynchronous nature in #20870 (comment).

I remember. Adding a task for monitoring PGs would be an improving worth adding in a new PR.

> Most pool operations rely on the send_command operation, which is truly asynchronous (no thread wrapping is required), and the tasks could leverage this, instead of wrapping an already asynchronous command inside new a thread. But in this PR we are not taking advantage of the asynchronous implementation of send_command.

Indeed. The TaskQueue in openATTIC is especially designed to simplify the handling of inherently asynchronous tasks. The TaskManager now creates new threads per default. 
> 
> Another detail that I think could be addressed in this PR is the fact that the pool creation operation is not waiting for the PGs to became clean as we also discussed previously.

There is IMO no need to wait for all PGs in order finish the creation of Pools: as soon as a Pool is created, you can work with it in the dashboard. Adding a new Task for showing the PG creation progress to the user would be nice extension.

> 
> In https://github.com/rjfd/ceph/blob/wip-dashboard-pr-20865/src/pybind/mgr/dashboard/controllers/pool.py you can find an example of the pool creation operation that takes advantage of the asynchronous send_command operation, and also waits for the PGs stabilization. You can use this example to as the basis for the implementation of the remaining pool operations.

The complexity of this code outweighs the benefit of not having a separate thread for a few seconds.

**Comment by sebastian-philipp** [comment]
Teuthology failed: 
```
======================================================================
FAIL: test_list (tasks.mgr.dashboard.test_rbd.RbdTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-tdehler-testing-21881-22757/qa/tasks/mgr/dashboard/test_rbd.py", line 214, in test_list
    self.assertEqual(len(data), 2)
AssertionError: 1 != 2

----------------------------------------------------------------------
Ran 79 tests in 783.126s

FAILED (failures=1)
```

**Comment by s0nea** [comment]
- https://shaman.ceph.com/builds/ceph/wip-tdehler-testing-21881/
- http://pulpito.ceph.com/tdehler-2018-07-03_13:59:23-rados:mgr-wip-tdehler-testing-21881-distro-basic-smithi/

Teuthology failed:
```cluster [WRN] Health check failed: 1 pools have pg_num > pgp_num (SMALLER_PGP_NUM)" in cluster log```

**Comment by s0nea** [comment]
Unfortunately the last QA tests run failed because of ansible issues:

- https://shaman.ceph.com/builds/ceph/wip-tdehler-testing-21881/96a7bbf422badfd1c8826c07b89bb5447436499d/
- http://pulpito.ceph.com/tdehler-2018-07-26_07:33:39-rados:mgr-wip-tdehler-testing-21881-distro-basic-smithi/

**Comment by s0nea** [comment]
The last QA test run was successful:

- https://shaman.ceph.com/builds/ceph/wip-tdehler-testing-21881/eb4c56bd51d30baebe4209f222eb17ed0b1f717d/
- http://pulpito.ceph.com/tdehler-2018-07-30_06:50:46-rados:mgr-wip-tdehler-testing-21881-distro-basic-smithi/

**Pull request merged**